### DISCUSSION TITLE
Using 74HC595 Shift Registers in a matrix

### DISCUSSION THREAD

**Discussion by esoterictriangle** [tttapa/Control-Surface#594] [Q&A]
this... might be off base. I know just enough to get dangerous ideas.

I know using SIPO shift registers as part of a matrix can be done--where you write outputs to the register(s) and listen on one or more pins--and my (likely rough) understanding of Control Surface's button matrix implementation is that rows are all set as output while columns as input, so I thought I might be able to do this:

```c++
#include <Control_Surface.h>

USBMIDI_Interface midi;

const AddresssMatrix<8, 1> addresses {{
    {60},
    {61},
    {62},
    {63},
    {64},
    {65},
    {66},
    {67},
}};

ShiftRegisterOut<8> sreg {
    7, //data (DS or SER)
    8, //clock (SH_CP or SRCLK)
    9, //latch (ST_CP or RCLK)
    MSBFIRST
};

NoteButtonMatrix<8,1> keyboard {
    {sreg.pin(0), sreg.pin(1), sreg.pin(2), sreg.pin(3), sreg.pin(4), sreg.pin(5), sreg.pin(6), sreg.pin(7)}, //rows--output
    {6}, //column
    addresses,
    CHANNEL_1,
};

void setup() {
    Control_Surface.begin();
}

void loop(){
    Control_Surface.loop();
}
```

but that doesn't seem to be the case. When connected to my promicro I get 0 messages if the diodes are one direction, and if I swap the diodes around any button will send all notes at the same time. I tried enabling debugging to see if that shed any light on the matter, but no debug messages display on the serial monitor, so I may be doing that wrong too 🤷‍♂️

Is there an issue with my code (or maybe my wiring? I can take a pic of the breadboard or make a diagram), or is this just something the library's not able to do right now?

**Comment by tttapa** [comment]
The button matrix class switches the row pins between input and output mode, it doesn't switch from high to low, which is why you cannot use a shift register (as it is always in output mode), see lines 24 and 38:

https://github.com/tttapa/Control-Surface/blob/e6542665da8997fdf33aee5ff1e0b7383c7ebe00/src/AH/Hardware/ButtonMatrix.ipp#L23-L39

That's not to say that you cannot do this using a shift register, of course, but you'll have your own variant of the button matrix class. Keep in mind that you have to use diodes in this case, otherwise you will short something out when pressing multiple buttons (this is the main reason why the provided button matrix code does not support it).

**Reply by esoterictriangle** [reply]
Ah, I think I begin to understand. Would the adjustment be as simple as changing lines 24 and 38 to `digitalWrite(rowPins[row], LOW);` and `digitalWrite(rowPins[row], HIGH);` respectively as well as adjusting line 50 to make rowPin OUTPUT instead of INPUT, or is just using HIGH/LOW too different from the current method of using High impedance (which... I don't think I fully understand and am reading more about right now)?

re: resistors: I understand diodes should be between the register's pins and the switches to prevent shorting the register, but I'm not sure how that will impact the code. Does it matter whether diodes are running between column and switch or row and switch? (If I was wiring a basic matrix, using the library as-is would it only work if I had diodes on the columns?)

**Reply by tttapa** [reply]
> Ah, I think I begin to understand. Would the adjustment be as simple as changing lines 24 and 38 to digitalWrite(rowPins[row], LOW); and digitalWrite(rowPins[row], HIGH); respectively as well as adjusting line 50 to make rowPin OUTPUT instead of INPUT

Indeed.

> or is just using HIGH/LOW too different from the current method of using High impedance (which... I don't think I fully understand and am reading more about right now)?

Not entirely sure what you mean, the pins of a serial-to-parallel shift register are outputs, so you can only switch between HIGH and LOW, you cannot switch individual pins to high-impedance mode.

> I understand diodes should be between the register's pins and the switches to prevent shorting the register, but I'm not sure how that will impact the code.

I _think_ using a single diode for each switch should be sufficient, you don't need additional diodes directly on the shift register pins. Try to follow the flow of current on a schematic to verify this. If you don't have diodes in your matrix, you can add a diode on each row pin, pointing into the shift register (i.e. it can sink current, but not source current).  
The diodes don't impact the code.

> Does it matter whether diodes are running between column and switch or row and switch? (If I was wiring a basic matrix, using the library as-is would it only work if I had diodes on the columns?)

Yes it matters, you need to use them like this: https://tttapa.github.io/Control-Surface-doc/Doxygen/d4/d53/NoteButtonMatrix_8ino-example.html

**Reply by esoterictriangle** [reply]
> I think using a single diode for each switch should be sufficient, you don't need additional diodes directly on the shift register pins. Try to follow the flow of current on a schematic to verify this. If you don't have diodes in your matrix, you can add a diode on each row pin, pointing into the shift register (i.e. it can sink current, but not source current).

ah, I think we're on the same page here--when I said shift register pins I meant Q0-Q7, not the microcontroller pins connected to data/clock/latch. Diodes are all pointing into the register.

I swapped out the code I mentioned earlier and things are responding well now save one issue: each button sends messages for _all_ buttons. I will note I only have four buttons actually wired on my breadboard right now to save time, but I don't think that should be causing this issue?
(for reference: [my fork's version of ButtonMatrix.ipp with edits](https://github.com/esoterictriangle/Control-Surface/blob/sreg-matrix/src/AH/Hardware/ButtonMatrix.ipp))

this is what each button press/release looks like in the serial monitor (it's also "fun" to listen to it when hooked up to a DAW)
```
09:16:47.394 -> Note On          Channel: 1	Data 1: 0x3e	Data 2: 0x7f
09:16:47.394 -> Note On          Channel: 1	Data 1: 0x3f	Data 2: 0x7f
09:16:47.394 -> Note On          Channel: 1	Data 1: 0x40	Data 2: 0x7f
09:16:47.394 -> Note On          Channel: 1	Data 1: 0x41	Data 2: 0x7f
09:16:47.394 -> Note On          Channel: 1	Data 1: 0x42	Data 2: 0x7f
09:16:47.394 -> Note On          Channel: 1	Data 1: 0x43	Data 2: 0x7f
09:16:47.394 -> Note On          Channel: 1	Data 1: 0x3c	Data 2: 0x7f
09:16:47.394 -> Note On          Channel: 1	Data 1: 0x3d	Data 2: 0x7f
09:16:47.579 -> Note Off         Channel: 1	Data 1: 0x42	Data 2: 0x7f
09:16:47.579 -> Note Off         Channel: 1	Data 1: 0x43	Data 2: 0x7f
09:16:47.626 -> Note Off         Channel: 1	Data 1: 0x3c	Data 2: 0x7f
09:16:47.626 -> Note Off         Channel: 1	Data 1: 0x3d	Data 2: 0x7f
09:16:47.626 -> Note Off         Channel: 1	Data 1: 0x3e	Data 2: 0x7f
09:16:47.626 -> Note Off         Channel: 1	Data 1: 0x3f	Data 2: 0x7f
09:16:47.626 -> Note Off         Channel: 1	Data 1: 0x40	Data 2: 0x7f
09:16:47.626 -> Note Off         Channel: 1	Data 1: 0x41	Data 2: 0x7f
```

**Reply by tttapa** [reply]
I don't have much time to look into it in detail, but you should probably write all row pins HIGH here:

https://github.com/esoterictriangle/Control-Surface/blob/d872476accf1c1102b4ca4aa1cbbe4bfc25940f5/src/AH/Hardware/ButtonMatrix.ipp#L48-L50

It probably won't solve your issue though, as it only affects the very first iteration.
